configserver-client/src/main/java/ai/vespa/hosted/client/AbstractConfigServerClient.java

```diff
@@ -0,0 +1,263 @@
+            return execute(this, requireNonNull(handler));
+        }
+
+        @Override
+        public <T> T read(Function<byte[], T> mapper) throws UncheckedIOException, ConfigServerException {
+            return mapIfSuccess(input -> {
+                try (input) {
+                    return mapper.apply(input.readAllBytes());
+                }
+                catch (IOException e) {
+                    throw new RetryException(e);
+                }
+            });
+        }
+
+        @Override
+        public void discard() throws UncheckedIOException, ConfigServerException {
+            mapIfSuccess(input -> {
+                try (input) {
+                    return null;
+                }
+                catch (IOException e) {
+                    throw new RetryException(e);
+                }
+            });
+        }
+
+        @Override
+        public InputStream stream() throws UncheckedIOException, ConfigServerException {
+            return mapIfSuccess(input -> input);
+        }
+
+        /** Returns the mapped body, if successful, retrying any IOException. The caller must close the body stream. */
+        private <T> T mapIfSuccess(Function<InputStream, T> mapper) {
+            return handle((response, ioException) -> {
+                if (response != null) {
+                    try {
+                        InputStream body = response.getEntity() != null ? response.getEntity().getContent()
+                                                                        : InputStream.nullInputStream();
+                        if (response.getCode() >= HttpStatus.SC_REDIRECTION)
+                            throw readException(body.readAllBytes());
+
+                        return mapper.apply(new ForwardingInputStream(body) {
+                            @Override
+                            public void close() throws IOException {
+                                super.close();
+                                response.close();
+                            }
+                        });
+                    }
+                    catch (IOException | RuntimeException | Error e) {
+                        try {
+                            response.close();
+                        }
+                        catch (IOException f) {
+                            e.addSuppressed(f);
+                        }
+                        if (e instanceof IOException)
+                            ioException = (IOException) e;
+                        else
+                            sneakyThrow(e);
+                    }
+                }
+                throw new RetryException(ioException);
+            });
+        }
+
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T extends Throwable> void sneakyThrow(Throwable t) throws T {
+        throw (T) t;
+    }
+
+    private static ConfigServerException readException(byte[] serialised) {
```

Comment:
This mechanism hides the HTTP response code, and the body unless it is in a specific format, and both are very useful.  E.g. the Orchestrator contains a detailed response on 409 Conflict.  I'm guessing Jetty may also return useful error responses (?).

Also, deserializing HTTP error responses in this class seems inconsistent with requiring clients to deserialize successful responses.

Reply:
This is what's currently done in the controller, and covers all uses there. I'll see whether it makes sense to change the orchestrator responses to match this, or not. 

So, this is what you have the `handle(...)` for. But you want both a different exception deserialisation _and_ easy response mapping at the same time, I guess. Fair enough. I'll get back to this.

Reply:
So there are three things: 
1. Handle successful response.
2. Handle non-2XX.
3. Handle exception.

Perhaps `handle` should accept a function for each of these, and overloads could let you specify only a subset. There's some code to handle exceptions and re-throw that the user shouldn't have to reimplement.

Reply:
Perhaps it should be possible to provide a function that overrides the default handling of (2), and:
 - if it throws an exception propagate that
 - if it returns true, fall back to default handling of (2)
 - if it returns false, fall back to handling as-if (1)

Similarly with (3).

An Orchestrator client could set custom (2) and (3).

Reply:
I think I'll rather expose a default handler for each, and make it easy to compose with those. But I started doubting whether a good default for 2. exists.

Reply:
We could just defer any such refactoring then until we know more about the config server clients.

Reply:
Yes, that's the plan :)